### PR TITLE
Better abstraction & https://golang.org/doc/effective_go.html#commentary

### PR DESCRIPTION
**Pull request by programaths**
# QA Round
- Added an interface and made _internal_ client unreachable from outside.
- Modified a bit the comments to comply with https://golang.org/doc/effective_go.html#commentary
- Used longer names for variables (clarity)

**Notes:**
Upgrade path : "SetToken" should be used to set the token instead of setting it directly on the client

So, this:

``` go
cli:=footballdata.NewClient(http.DefaultClient)
cli.AuthToken="abcd"
```

Becomes this:

``` go
cli:=footballdata.NewClient(http.DefaultClient)
cli.SetToken("abcd")
```

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -15,24 +15,52 @@ Provides a high-level client to talk to the API that football-data.org offers.
 
 To create an instance please use NewClient(h).
 */
-type Client struct {
-	h *http.Client
+type Client interface {
```
Comment (programaths) [inline client.go:18]: Interface offered to users

File: client.go [hunk 2]
```diff
@@ -15,24 +15,52 @@ Provides a high-level client to talk to the API that football-data.org offers.
...
+
+	SetToken(authToken string)
+}
+
+/*
+Provides a high-level client implementation to talk to the API that football-data.org offers.
+
+To create an instance please use NewClient(h).
+*/
+type client struct {
```
Comment (programaths) [inline client.go:37]: Not anymore instantiable from the outside. Hence, factory function is garanted to be the sole entry point.

File: client.go [hunk 3]
```diff
@@ -15,24 +15,52 @@ Provides a high-level client to talk to the API that football-data.org offers.
...
+//
+// Call SetToken to add your token.
+func NewClient(h *http.Client) Client {
+	return &client{httpClient: h}
 }
 
-func (c *Client) req(path string, pathValues ...interface{}) request {
+// SetToken Set the authentication token
+// Calling this method is *optional*
+func (c* client) SetToken(authToken string) {
```
Comment (programaths) [inline client.go:53]: This is now required to give write only rights to user land code

File: client.go [hunk 4]
```diff
@@ -48,7 +76,7 @@ func (c *Client) doJson(method string, path string, values url.Values) (j *json.
 	req.Header.Set("User-Agent", "go-footballdata/0.0")
 
 	// Execute request
-	resp, err := c.h.Do(req)
+	resp, err := c.httpClient.Do(req)
```
Comment (programaths) [inline client.go:79]: Long name for clarity

File: req_fixture.go [hunk 5]
```diff
@@ -4,13 +4,13 @@ import "fmt"
 
 type FixtureRequest struct{ request }
 
-// Modifies the request to specify the number of former games to be analyzed (normally 10).
+// Head2Head Modifies the request to specify the number of former games to be analyzed (normally 10).
```
Comment (programaths) [inline req_fixture.go:7]: See https://golang.org/doc/effective_go.html#commentary

Comment (icedream) [inline req_fixture.go:7]: I guess "modifies" should be written lowercase here - other than that I think I will apply this generally for all methods with the next commits.

**Comment by icedream** [comment]
Awesome pull request. Thanks for the contribution!

Since there's no tagged release of the implementation yet I don't think any of the changes need to be explicitly documented yet. I'll merge this as is for now and see if I can add some more changes to complement this pull request.

**Pull request merged**